Make legend appearance consistent with chart elements

nagix opened the pull request:
### Problems

Legend appearance is not consistent with chart elements. For example,

1. In a bar chart, the border of the legend is rendered while bars have no border
2. In a bar chart, the border width of the legend is 3 pixels while the width of bars is 6 pixels
3. In a bar chart, the border width and color of the legend don't match any of bars
4. In a line chart, the colors and border width of the legend don't match those of points when `usePointStyle` is `true` (#5375)
5. In a bar chart, a dashed border is used in the legend while bars have no dashed border
6. In a line chart, the rotation of the point in the legend doesn't match the one in the chart when `usePointStyle` is `true` and `pointRotation` is non-zero.
7. In a line chart, the legend box is filled even if `fill` is `false`.

See https://jsfiddle.net/nagix/5ztb6hLc

<img width="434" alt="screen shot 2018-07-07 at 4 48 27 am" src="https://user-images.githubusercontent.com/723188/42399888-1d3c544e-81a2-11e8-8711-5948bb9afcd1.png">
<img width="434" alt="screen shot 2019-02-13 at 12 11 26 am" src="https://user-images.githubusercontent.com/723188/52649888-248c1880-2f24-11e9-983f-aa764dbe7bdf.png">

### Solutions

1. The default value for line elements was always used. Use the default value for each element type if a property is not set.
2. A border was rendered before a background was rendered. Stroke after fill to move borders to the front.
3. The border width and color of the first element should be picked up. Support indexable and scriptable options.
4. Use point's colors and border width when `usePointStyle` is `true`.
5. Disable line styles (dash, cap and join) when `usePointStyle` is `true` or the dataset type is not `'line'` or `'radar'`.
6. Use point's rotation when `usePointStyle` is `true`.
7. The background color is set to transparent when `fill` is `false`.

See https://jsfiddle.net/nagix/negp2rLh

<img width="434" alt="screen shot 2018-07-07 at 4 50 12 am" src="https://user-images.githubusercontent.com/723188/42400171-23146f18-81a3-11e8-95c9-a524dd06b7dc.png">
<img width="434" alt="screen shot 2019-02-13 at 12 12 02 am" src="https://user-images.githubusercontent.com/723188/52649901-29e96300-2f24-11e9-91fb-2e034d3a5e3e.png">

In addition to above, the existing tests are fixed and more tests are added.

Fixes #5375
Fixes #6135
Closes #2281

benmccann commented on src/plugins/plugin.legend.js:
```diff
@@ -47,17 +47,36 @@ defaults._set('global', {
 					return {
 						text: dataset.label,
-						fillStyle: (!helpers.isArray(dataset.backgroundColor) ? dataset.backgroundColor : dataset.backgroundColor[0]),
+						fillStyle: valueAtIndexOrDefault(backgroundColor, 0, elementOpts.backgroundColor),
```
Are we trying to prefer `resolve` to `valueAtIndexOrDefault` now after https://github.com/chartjs/Chart.js/pull/5965? Or I think at least we want to make those variables global instead of in the function?

nagix replied:
Thanks for reminding me. Since this involves the checks for `usePointStyle`, I would keep using `valueAtIndexOrDefault` and just make it global.

etimberg commented on test/specs/plugin.legend.tests.js:
```diff
@@ -565,6 +625,107 @@ describe('Legend block tests', function() {
 		});
 	});
 
+	it('should pick up the first item when the propertiy is an array', function() {
```
[nit] typo in 'property'

nagix replied:
Done

kurkle commented:
To me this seems ready to merge, @etimberg, @simonbrunel ?

nagix commented:
Recently scriptable options have been added, so I will review the code.

nagix commented:
Added support for scriptable options and `pointRotation/rotation`.

simonbrunel commented on src/controllers/controller.doughnut.js:
```diff
@@ -45,20 +45,14 @@ defaults._set('doughnut', {
-						var fill = resolve([custom.backgroundColor, ds.backgroundColor, arcOpts.backgroundColor], undefined, i);
-						var stroke = resolve([custom.borderColor, ds.borderColor, arcOpts.borderColor], undefined, i);
-						var bw = resolve([custom.borderWidth, ds.borderWidth, arcOpts.borderWidth], undefined, i);
+						var model = meta.controller._resolveElementOptions(meta.data[i] || {}, i);
```
Should we use `var model = meta.data[i]._options` instead? (as [defined here](/chartjs/Chart.js/blob/master/src/controllers/controller.doughnut.js#L177))
Or maybe `var model = meta.data[i]._model`?

nagix replied:
There is a change that a dataset has no data but styles. `_model` cannot be used if there is no data in the dataset.

simonbrunel replied:
You right

simonbrunel commented on src/plugins/plugin.legend.js:
```diff
@@ -48,17 +48,35 @@ defaults._set('global', {
+					var controller = meta.controller;
+					var legendOpts = chart.options.legend;
+					var usePointStyle = legendOpts && legendOpts.labels && legendOpts.labels.usePointStyle;
+					var model;
```
Could be wrong but does it make sense to use the already computed `_model` instead?

```javascript
var element = meta.dataset || (meta.data && meta.data[0]);
var model = element ? element._model : {};
```

nagix replied:
Same as above. `_model` cannot be used if there is no data in the dataset. Maybe we can use `_model` for lines only.

simonbrunel replied:
`_resolve*Options` are specific and internal to our built-in controllers and should not be expected for external implementations, so I think this change will break some integrations.

simonbrunel commented:
@nagix I think there is a deeper issue with the current implementation (master) because the legend needs to know about the dataset options logic, which is specific to the controller. I also don't think controllers should be responsible to override `generateLabels`, neither have the knowledge of the legend plugin (same issues with [the tooltip](/chartjs/Chart.js/blob/master/src/core/core.tooltip.js#L80) labels).

Thinking out loud: what about introducing a new dataset API to generate a "style", so dataset controllers would be able to return a set of predefined style properties that should be used to represent the dataset.

For example (line controller):

```javascript
// Part of the line dataset controller
// @params {string} type - "primary" / "secondary"
// @return {IStyleInterface}
getStyle: function(type) {
    var options = type === "primary"
       ? this._resolveLineOptions(...)
       : this._resolvePointOptions(...);

    return {
        backgroundColor: options.backgroundColor,
        borderColor: options.borderColor,
        // ...
    };
}
```

For backward compatibility, we will need a base implementation at the DatasetController level. Then the legend default `generateLabels` would be something like:

```javascript
generateLabels: function(chart) {
    var options = chart.options.legend || {};
    var secondary = options.labels && legendOpts.labels.usePointStyle;
    var type = secondary ? 'secondary' : 'primary';

    return helpers.isArray(data.datasets) ? data.datasets.map(function(dataset, i) {
        var meta = chart.getDatasetMeta(i);
        var controller = meta.controller;
        var style = controller.getStyle(type);

        return {
            text: dataset.label,
            fillStyle: style.backgroundColor,
            // ...
        };
    } : [];
}
```

We could use the same call in the tooltip and certainly and get rid of the custom `generateLabels` defaults in the radar and polar are controllers. Of course the suggested API is an example and would need to be discussed (`getStyle`, `IStyleInterface`, etc.) but you get the idea.

What do you think about this approach?

nagix commented:
@simonbrunel I like this approach. `getStyle` could be used for other purposes like annotations and more. Let me write code to see how it works. Do you think the doughnut and polarArea controller still need to override `generateLabels` to support per-element labels?

simonbrunel commented:
@nagix I forgot about the per-element legend labels, so that means we need to support accessing style for the dataset or for a specific data. Maybe `getStyle(index)`, with `index` the index of the data to retrieve the style or, if `index: null/undefined`, the style of the dataset (so no need of the `type` previously suggested).

I'm not sure we can remove the overrides in the doughnut and polar controllers yet. Maybe we could add a new legend options to switch between per-dataset / per-data legend items (e.g. `source: 'dataset' | 'data'`), but that would be part of another PR. Though, the current per-data approach is a bit strange in case of multiple datasets with different styles.

kurkle commented on src/controllers/controller.radar.js:
```diff
@@ -102,6 +102,16 @@ module.exports = DatasetController.extend({
+	 * @private
+	 */
+	_resolveElementOptions: function(element, index) {
+		if (index !== undefined) {
```
[nit] Writing like this should save couple of bytes:
```js
return index >= 0
    ? this._resolvePointOptions(element, index)
    : this._resolveLineOptions(element);
```

nagix replied:
Done

kurkle commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,49 @@ helpers.extend(DatasetController.prototype, {
+			style.backgroundColor = 'rgba(0,0,0,0)';
+		}
+
+		return {
```
I think its safe to just `return style` here?

nagix replied:
I explicitly list the properties because we are using `IStyleInterface` as a return value.

kurkle replied:
Where is this `IStyleInterface` defined and why should we care in javascript?

nagix replied:
It’s not defined in Chart.js code and I thought some TypeScript code may use it. I was not quite sure if we need to care this in JS code.

nagix replied:
Made the suggested change

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -46,20 +46,14 @@ defaults._set('doughnut', {
-						var fill = resolve([custom.backgroundColor, ds.backgroundColor, arcOpts.backgroundColor], undefined, i);
-						var stroke = resolve([custom.borderColor, ds.borderColor, arcOpts.borderColor], undefined, i);
-						var bw = resolve([custom.borderWidth, ds.borderWidth, arcOpts.borderWidth], undefined, i);
+						var style = meta.controller.getStyle(i);
```
I wonder what the performance impact of this will be. Before we were only resolving three properties and now we're resolving seven

I also noticed that in `_resolveElementOptions` we're resolving the hover properties, but then I don't think we're ever using them. `setHoverStyle` seems to do the resolution again and ignores the previously resolved styles. Perhaps we should not be trying to resolve them in `_resolveElementOptions`

nagix replied:
Maybe we can use cached `element._options` if the element exists since element updates are done prior to layout updates. Some controllers are currently not using `_options`, so we need to update them.

Regarding `setHoverStyle`, most of controllers other than the bar controller are using cached styles, and resolving on update is more efficient than doing on hover event. Further optimization may be possible, though (eg. `valueOrDefault(options.hoverBackgroundColor, getHoverColor(options.backgroundColor))` can be done in `_resolveElementOptions`).

nagix replied:
Updated the code to use `element._options` if available.

benmccann commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,49 @@ helpers.extend(DatasetController.prototype, {
 		}
 	},
 
+	/*
```
I believe this should start with `/**`

nagix replied:
Done

benmccann commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,49 @@ helpers.extend(DatasetController.prototype, {
+	 * @private
+	 */
+	_resolveElementOptions: function() {
+		// noop
```
It seems like this should throw an `Error` letting users know that they need to implement this method. If this method is not implemented then `getStyle` will be broken and it will be harder to track down why vs. if this threw an `Error`

nagix replied:
An error message was added. `_resolveElementOptions` has a lot of duplication between controllers, so probably we need further refactoring (in other PR).

etimberg commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,49 @@ helpers.extend(DatasetController.prototype, {
+	 * @params {number} index - data index
+	 * @return {IStyleInterface} style object
+	 */
+	getStyle: function(index) {
```
I like the idea of creating a function here for this. Perhaps eventually the tooltip could use this too

Hedva commented:
Thanks for putting time and effort in this project guys. Much appreciated. 
Can´t wait for this feature to be released

nagix commented:
I reverted the last change because I realized that the layout updates are done prior to the element updates, and old style options may be picked.

simonbrunel commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,39 @@ helpers.extend(DatasetController.prototype, {
+	 * @private
+	 */
+	_resolveElementOptions: function() {
+		throw new Error('This method is not implemented.');
```
Isn't it a breaking change for custom (external) controllers inheriting the dataset controller directly, thus would not have this method implemented? Should we provide a base implementation that would resolve only common options?

nagix replied:
Implemented this in the datasetController.

simonbrunel commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,39 @@ helpers.extend(DatasetController.prototype, {
+		}
+		style = me._resolveElementOptions(element || {}, index);
+
+		if (style.fill === false || style.fill === null) {
```
Why do we need this case?

nagix replied:
This is for addressing the [problem 7](https://github.com/chartjs/Chart.js/pull/5621#issue-199847518).

simonbrunel commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +268,39 @@ helpers.extend(DatasetController.prototype, {
+	 * @params {number} index - data index
+	 * @return {IStyleInterface} style object
+	 */
+	getStyle: function(index) {
```
@nagix I know it comes from [my suggestion](/chartjs/Chart.js/pull/5621#issuecomment-463627999) but do you think it really needs to be public? I'm fine with it, just want your thought (@etimberg @kurkle @benmccann?)

benmccann replied:
I probably would have gone with making it private since it's new just to give any issues a chance to shake out, but don't feel strongly

kurkle replied:
I think its fine being public.

nagix commented:
Implemented `_resolveElementOptions` in the base datasetController and refactored derived controllers to eliminate duplication between controllers. Now each derived controller has `_optionKey` property, and element options will be resolved by `_resolveElementOptions` in the base datasetController.

benmccann commented on src/controllers/controller.bubble.js:
```diff
@@ -122,16 +138,13 @@ module.exports = DatasetController.extend({
 	 * @private
 	 */
-	_resolveElementOptions: function(point, index) {
+	_resolveDataElementOptions: function(point, index) {
```
why add `Data` to the name? I liked the name `_resolveElementOptions`

nagix replied:
Now there are two resolving functions in the base controller: `_resolveDatasetElementOptions` and `_resolveDataElementOptions`. The former resolves dataset element options (i.e. line element options) and the latter resolves data element options (arc, point and rectangle element options). I wanted to make it clear that each method resolves which element options.

nagix replied:
Since some controllers have custom resolving logic and extend resolving methods, I think it is better to have these methods separately.

benmccann replied:
Ah, the names `_resolveDataElementOptions` and `_resolveDatasetElementOptions` are so similar I did not notice there were two different methods

kurkle replied:
(I edited @nagix's comment, changed 2nd `_datasetElementOptions` to `_dataElementOptions`)

nagix replied:
Thanks @kurkle

simonbrunel commented:
@kurkle I'm not sure options should be at the element level, it would give too much responsibility to these classes (i.e. how options should be exposed to the public API) but also make the whole thing less flexible. I like to have each controller deciding about their own options API.

kurkle commented:
@simonbrunel I don't see how it would affect flexibility. It could bring new options to a controller that it didn't know of. For example add gradient fill to arc element, doughnut and polarArea would immediately support those through their options (not just through `elements.arc` that it should obey anyway, shouldn't it?)

It would also help consistency in option names and make them easier to document.

Just my opinion and I'm not going to be pushing it (more).

nagix commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +297,110 @@ helpers.extend(DatasetController.prototype, {
+	/**
+	 * Returns a set of predefined style properties that should be used to represent the dataset
+	 * or the data if the index is specified
+	 * @params {number} index - data index
```
Typo fixed

benmccann commented on src/core/core.datasetController.js:
```diff
@@ -268,6 +297,110 @@ helpers.extend(DatasetController.prototype, {
+	/**
+	 * @private
+	 */
+	_resolveDataElementOptions: function(element, index) {
```
Optimization idea: if none of the option values are arrays or functions then cache the resolved options and return the same result for every index

benmccann commented:
@nagix this PR will need to be rebased

simonbrunel commented:
Thanks @nagix for this work (and your patience) :)
